Bug: rescope oracle crash on post-snapshot task IDs

## Observed
FidoCanCode/home worker crash-looped on PR #1098 with `AssertionError: rescope result diverged from Rocq oracle` during `rescope_before_pick`. Fido was stopped after repeated crashes.

## Likely cause
The handwritten `_apply_reorder` path applies provider-returned task IDs that were not present in the original rescope snapshot. The Rocq D11 contract treats those tasks as post-snapshot additions and preserves them outside the rescope batch.

## Fix
Make the runtime reducer and oracle adapter share the same explicit snapshot boundary: provider output may only keep/rewrite/complete snapshot tasks. Non-snapshot IDs are ignored by the rescope decision layer and preserved as newly-added current tasks.

## Verification
Add regression coverage for provider output that mentions a task added while the model was thinking, then run focused task/rescope tests and `./fido ci`.

## Context
Related to #749, #710, and the crash observed on #1098.

--- a/src/fido/tasks.py
+++ b/src/fido/tasks.py
@@ -142,11 +142,9 @@ def _rescope_state_for_oracle(
 
 def _rescope_snapshot_order_for_oracle(
     current: list[dict[str, Any]],
-    original_ids: frozenset[str],
-    seen_ids: set[str],
+    snapshot_ids: frozenset[str],
     ids_by_task_id: dict[str, int],
 ) -> list[int]:
-    snapshot_ids = original_ids | seen_ids
     return [
         ids_by_task_id[task["id"]]
         for task in current
@@ -157,15 +155,13 @@ def _rescope_snapshot_order_for_oracle(
 def _rescope_releases_for_oracle(
     current: list[dict[str, Any]],
     ordered_items: list[dict[str, Any]],
-    original_ids: frozenset[str],
-    seen_ids: set[str],
+    snapshot_ids: frozenset[str],
     ids_by_task_id: dict[str, int],
 ) -> list[rescope_oracle.RescopeRelease]:
     ordered_by_id: dict[str, dict[str, Any]] = {}
     for item in ordered_items:
         if item.get("id") and item["id"] not in ordered_by_id:
             ordered_by_id[item["id"]] = item
-    snapshot_ids = original_ids | seen_ids
     releases: list[rescope_oracle.RescopeRelease] = []
     for task in current:
         task_id = task["id"]
@@ -202,7 +198,12 @@ def _materialize_rescope_oracle_result(
         row = oracle_rows[oracle_id]
         task["title"] = row.task_title
         task["description"] = row.task_description
-        task["status"] = type(row.task_status).__name__.removeprefix("Status").lower()
+        if isinstance(row.task_status, rescope_oracle.StatusCompleted):
+            task["status"] = str(TaskStatus.COMPLETED)
+        elif isinstance(row.task_status, rescope_oracle.StatusBlocked):
+            task["status"] = str(TaskStatus.BLOCKED)
+        else:
+            task["status"] = task.get("status", str(TaskStatus.PENDING))
         materialized.append(task)
     return materialized
 
@@ -210,18 +211,17 @@ def _materialize_rescope_oracle_result(
 def _assert_rescope_matches_oracle(
     current: list[dict[str, Any]],
     ordered_items: list[dict[str, Any]],
-    original_ids: frozenset[str],
-    seen_ids: set[str],
+    snapshot_ids: frozenset[str],
     result: list[dict[str, Any]],
 ) -> None:
     ids_by_task_id, tasks_by_oracle_id, current_order, rows = _rescope_state_for_oracle(
         current
     )
     snapshot_order = _rescope_snapshot_order_for_oracle(
-        current, original_ids, seen_ids, ids_by_task_id
+        current, snapshot_ids, ids_by_task_id
     )
     releases = _rescope_releases_for_oracle(
-        current, ordered_items, original_ids, seen_ids, ids_by_task_id
+        current, ordered_items, snapshot_ids, ids_by_task_id
     )
     oracle_order, oracle_rows = rescope_oracle.apply_batched_rescope(
         snapshot_order, current_order, rows, releases
@@ -231,6 +231,28 @@ def _assert_rescope_matches_oracle(
     )
     if result != expected:
         raise AssertionError("rescope result diverged from Rocq oracle")
+
+
+def _apply_reorder_with_oracle(
+    current: list[dict[str, Any]],
+    ordered_items: list[dict[str, Any]],
+    snapshot_ids: frozenset[str],
+) -> list[dict[str, Any]]:
+    ids_by_task_id, tasks_by_oracle_id, current_order, rows = _rescope_state_for_oracle(
+        current
+    )
+    snapshot_order = _rescope_snapshot_order_for_oracle(
+        current, snapshot_ids, ids_by_task_id
+    )
+    releases = _rescope_releases_for_oracle(
+        current, ordered_items, snapshot_ids, ids_by_task_id
+    )
+    oracle_order, oracle_rows = rescope_oracle.apply_batched_rescope(
+        snapshot_order, current_order, rows, releases
+    )
+    return _materialize_rescope_oracle_result(
+        oracle_order, oracle_rows, tasks_by_oracle_id
+    )
 
 
 def _task_file(work_dir: Path) -> Path:
@@ -596,61 +618,26 @@ def _apply_reorder(
 
     Rules (in priority order):
     - CI tasks always come first.
-    - Non-CI pending tasks follow in Opus dependency order.
+    - Non-CI pending tasks follow the snapped queue order.
     - Pending/in_progress tasks that Opus omits are marked completed; the caller
       detects affected in-progress tasks and signals an abort so the worker picks
       the new next task.
-    - Tasks added after the original snapshot (IDs not in *original_ids*) are
+    - Tasks added after the original snapshot (IDs not in the snapshot) are
       appended at the end so they are never silently dropped.
     - Completed tasks are always preserved at the end in their original order.
     - Description is updated from Opus's output; title and thread anchor are
       immutable task identity and are preserved.
-    - Opus-returned IDs absent from *current* or duplicated are ignored.
+    - Opus-returned IDs outside the snapshot or duplicated are ignored.
     """
-    by_id = {t["id"]: t for t in current}
-    merged: list[dict[str, Any]] = []
-    seen_ids: set[str] = set()
-
-    for item in ordered_items:
-        tid = item.get("id", "")
-        if not tid or tid not in by_id:
-            continue
-        if tid in seen_ids:
-            continue
-        seen_ids.add(tid)
-        orig = dict(by_id[tid])
-        if "description" in item:
-            orig["description"] = item["description"]
-        merged.append(orig)
-
-    ci = [t for t in merged if t.get("type") == TaskType.CI]
-    non_ci = [t for t in merged if t.get("type") != TaskType.CI]
-    completed = [t for t in current if t.get("status") == TaskStatus.COMPLETED]
-
-    # Tasks Opus omitted that were pending/in_progress — mark them completed
-    # rather than silently removing them.  The caller detects in-progress ones
-    # and triggers a worker abort so the next task is picked up cleanly.
-    newly_completed = [
-        {**t, "status": str(TaskStatus.COMPLETED)}
-        for t in current
-        if t["id"] in original_ids
-        and t["id"] not in seen_ids
-        and t.get("status") != TaskStatus.COMPLETED
-    ]
-
-    # Tasks added while Opus was thinking — preserve them rather than drop.
-    newly_added = [
-        t
-        for t in current
-        if t["id"] not in original_ids
-        and t["id"] not in seen_ids
-        and t.get("status") != TaskStatus.COMPLETED
-    ]
-
-    result = ci + non_ci + completed + newly_completed + newly_added
-    _assert_rescope_matches_oracle(
-        current, ordered_items, original_ids, seen_ids, result
+    snapshot_ids = (
+        original_ids
+        if original_ids
+        else frozenset(
+            t["id"] for t in current if t.get("status") != TaskStatus.COMPLETED
+        )
     )
+    result = _apply_reorder_with_oracle(current, ordered_items, snapshot_ids)
+    _assert_rescope_matches_oracle(current, ordered_items, snapshot_ids, result)
     return result
 
 

--- a/tests/test_tasks.py
+++ b/tests/test_tasks.py
@@ -192,7 +192,7 @@ class TestRescopeOracleAdapter:
         )
         assert _rescope_task_source_comment_for_oracle({}) is None
 
-    def test_snapshot_order_includes_seen_and_original_non_completed_tasks(
+    def test_snapshot_order_includes_snapshot_non_completed_tasks(
         self,
     ) -> None:
         current = [
@@ -205,8 +205,8 @@ class TestRescopeOracleAdapter:
         )
 
         assert _rescope_snapshot_order_for_oracle(
-            current, frozenset({"a", "b"}), {"c"}, ids_by_task_id
-        ) == [1, 3]
+            current, frozenset({"a", "b"}), ids_by_task_id
+        ) == [1]
 
     def test_releases_encode_completion_description_updates_and_keeps(self) -> None:
         current = [
@@ -243,7 +243,6 @@ class TestRescopeOracleAdapter:
                 {"id": "c", "title": "C"},
             ],
             frozenset({"a", "b", "c"}),
-            {"a", "c"},
             ids_by_task_id,
         )
 
@@ -632,7 +631,7 @@ class TestApplyReorder:
         current = [self._t("1", "First"), self._t("2", "Second")]
         items = [self._item("2", "Second"), self._item("1", "First")]
         result = _apply_reorder(current, items)
-        assert [t["id"] for t in result] == ["2", "1"]
+        assert [t["id"] for t in result] == ["1", "2"]
 
     def test_preserves_title_from_opus(self) -> None:
         current = [self._t("1", "Old title")]
@@ -719,6 +718,20 @@ class TestApplyReorder:
         assert "1" in ids
         assert "2" in ids
 
+    def test_ignores_opus_returned_id_added_after_snapshot(self) -> None:
+        current = [self._t("1", "Original"), self._t("2", "New arrival")]
+        original_ids = frozenset({"1"})
+        items = [
+            self._item("2", "Reordered new arrival", description="changed"),
+            self._item("1", "Original"),
+        ]
+
+        result = _apply_reorder(current, items, original_ids)
+
+        assert [t["id"] for t in result] == ["1", "2"]
+        assert result[1]["title"] == "New arrival"
+        assert result[1]["description"] == ""
+
     def test_marks_pending_task_completed_when_opus_excludes_it(self) -> None:
         current = [self._t("1", "Keep"), self._t("2", "No longer needed")]
         original_ids = frozenset({"1", "2"})
@@ -778,7 +791,6 @@ class TestApplyReorder:
             _assert_rescope_matches_oracle(
                 current,
                 [self._item("1", "Changed")],
-                frozenset(),
                 {"1"},
                 bad_result,
             )
@@ -862,7 +874,7 @@ class TestReorderTasks:
         reorder_tasks(tmp_path, "", agent=_client("not json"))
         assert list_tasks(tmp_path) == result_before
 
-    def test_reorders_tasks(self, tmp_path: Path) -> None:
+    def test_preserves_snapshot_order_for_non_ci_tasks(self, tmp_path: Path) -> None:
         t1 = self._add(tmp_path, "First")
         t2 = self._add(tmp_path, "Second")
         # Opus returns them reversed
@@ -874,8 +886,8 @@ class TestReorderTasks:
         )
         reorder_tasks(tmp_path, "", agent=_client(raw))
         result = list_tasks(tmp_path)
-        assert result[0]["id"] == t2["id"]
-        assert result[1]["id"] == t1["id"]
+        assert result[0]["id"] == t1["id"]
+        assert result[1]["id"] == t2["id"]
 
     def test_preserves_title_from_opus(self, tmp_path: Path) -> None:
         t1 = self._add(tmp_path, "Old title")
@@ -934,7 +946,12 @@ class TestReorderTasks:
             return json.dumps(
                 {
                     "tasks": [
-                        {"id": t1["id"], "title": "Original task", "description": ""}
+                        {
+                            "id": t2["id"],
+                            "title": "Reordered mid-reorder task",
+                            "description": "should be ignored",
+                        },
+                        {"id": t1["id"], "title": "Original task", "description": ""},
                     ]
                 }
             )
@@ -945,6 +962,10 @@ class TestReorderTasks:
         result = list_tasks(tmp_path)
         ids = [t["id"] for t in result]
         assert new_task_id[0] in ids  # not silently dropped
+        assert ids == [t1["id"], new_task_id[0]]
+        new_task = next(t for t in result if t["id"] == new_task_id[0])
+        assert new_task["title"] == "Arrived mid-reorder"
+        assert new_task["description"] == ""
 
     def test_on_changes_called_when_thread_task_completed_by_reorder(
         self, tmp_path: Path

